encoder.go

```diff
@@ -38,67 +46,93 @@ func newMaxEncoder() *encoder {
 	}
 }
 
+// Encode takes a Go value and returns a wafObject pointer and an error.
+// The returned wafObject is the root of the tree of imbriqued wafObjects representing the Go value.
+// The only error case is if the top-level object is "Unusable" which means that the data is nil or a non-data type
+// like a function or a channel.
 func (encoder *encoder) Encode(data any) (*wafObject, error) {
 	value := reflect.ValueOf(data)
 	wo := &wafObject{}
 
-	if err := encoder.encode(value, wo, encoder.objectMaxDepth); err != nil {
-		return nil, err
+	encoder.encode(value, wo, encoder.objectMaxDepth)
+
+	if wo._type == wafInvalidType {
+		return nil, errUnsupportedValue
 	}
 
 	return wo, nil
 }
 
-func encodeNative[T native](val T, t wafObjectType, obj *wafObject) error {
+func encodeNative[T native](val T, t wafObjectType, obj *wafObject) {
 	obj._type = t
 	obj.value = (uintptr)(val)
-	return nil
 }
 
-func (encoder *encoder) encode(value reflect.Value, obj *wafObject, depth int) error {
+var nullableTypeKinds = map[reflect.Kind]struct{}{
+	reflect.Interface:     {},
+	reflect.Pointer:       {},
+	reflect.UnsafePointer: {},
+	reflect.Map:           {},
+	reflect.Slice:         {},
+	reflect.Func:          {},
+	reflect.Chan:          {},
+}
+
+// isValueNil check if the value is nullable and if it is actually nil
+// we cannot directly use value.IsNil() because it panics on non-pointer values
+func isValueNil(value reflect.Value) bool {
+	_, nullable := nullableTypeKinds[value.Kind()]
+	return nullable && value.IsNil()
+}
+
+func (encoder *encoder) encode(value reflect.Value, obj *wafObject, depth int) {
+	value.IsValid()
 	switch kind := value.Kind(); {
 	// Terminal cases (leafs of the tree)
 	case kind == reflect.Invalid:
-		return errUnsupportedValue
+		encodeNative[uintptr](0, wafInvalidType, obj)
```

Comment:
Alright I guess I see why you can't infer the `obj._type` from the type argument... Somehow I dislike this particular calling pattern, but I guess... fine.

Reply:
right

Reply:
refactored this